File: packages/types/schema/vehicle.ts
import { z } from 'zod';

// 차량 상태 열거형 - 통합된 정의
export enum _VehicleStatusEnum {
  ACTIVE = 'active',
  AVAILABLE = 'available',
  IN_USE = 'in_use',
  MAINTENANCE = 'maintenance',
  INACTIVE = 'inactive',
  RESERVED = 'reserved',
  OUT_OF_SERVICE = 'out_of_service',
}

// Zod 스키마용 VehicleStatus
export const VehicleStatusSchema = z.nativeEnum(_VehicleStatusEnum);

// 차량 유형 열거형
export enum VehicleType {
  SEDAN = 'sedan',
  SUV = 'suv',
  TRUCK = 'truck',
  VAN = 'van',
  MOTORCYCLE = 'motorcycle',
  BUS = 'bus',
  OTHER = 'other',
}

// 차량 Zod 스키마 정의
export const VehicleSchema = z.object({
  id: z.string().uuid(),
  organizationId: z.string().uuid().optional(),
  licensePlate: z.string(),
  plateNumber: z.string().optional(), // 기존 코드 호환성을 위해 추가
  make: z.string(),
  model: z.string(),
  year: z
    .number()
    .int()
    .min(1900)
    .max(new Date().getFullYear() + 1),
  vehicleType: z.nativeEnum(VehicleType).optional(),
  vin: z.string().optional(),
  color: z.string().optional(),
  status: VehicleStatusSchema,
  mileage: z.number().int(),
  fuelType: z.string().optional(),
  transmission: z.string().optional(),
  engine: z.string().optional(),
  features: z.array(z.string()).optional(),
  notes: z.string().optional(),
  ownerId: z.string().uuid().optional(),
  registrationDate: z.date().optional(),
  createdAt: z.date(),
  updatedAt: z.date(),
});

// 단일 Vehicle 타입 정의
export type Vehicle = z.infer<typeof VehicleSchema>;

// 차량 생성 입력값 스키마
export const CreateVehicleInputSchema = VehicleSchema.omit({
  id: true,
  createdAt: true,
  updatedAt: true,
});

export type CreateVehicleInput = z.infer<typeof CreateVehicleInputSchema>;

// 차량 업데이트 입력값 스키마
export const UpdateVehicleInputSchema = VehicleSchema.omit({
  id: true,
  createdAt: true,
  updatedAt: true,
}).partial();

export type UpdateVehicleInput = z.infer<typeof UpdateVehicleInputSchema>;

// 차량 생성 DTO 타입 (기존 호환성 유지)
export type CreateVehicleDto = Omit<Vehicle, 'id' | 'status' | 'createdAt' | 'updatedAt'>;

// 차량 업데이트 DTO 타입 (기존 호환성 유지)
export type UpdateVehicleDto = Partial<Omit<Vehicle, 'id' | 'createdAt' | 'updatedAt'>>;

// 차량 위치 상태 열거형
export enum LocationStatus {
  ACTIVE = 'active',
  IDLE = 'idle',
  MAINTENANCE = 'maintenance',
  OUT_OF_SERVICE = 'out_of_service',
  OFFLINE = 'offline',
}

// 차량 위치 타입
export interface VehicleLocation {
  id: string;
  vehicleId: string;
  licensePlate?: string;
  latitude: number;
  longitude: number;
  altitude?: number;
  heading?: number;
  speed?: number;
  accuracy?: number;
  status: 'active' | 'idle' | 'maintenance' | 'out_of_service' | 'offline';
  timestamp: number;
  createdAt: string;
  updatedAt: string;
}

// 차량 위치 생성 DTO 타입
export type CreateVehicleLocationDto = Omit<
  VehicleLocation,
  'id' | 'licensePlate' | 'timestamp' | 'createdAt' | 'updatedAt'
> & { timestamp?: number };

// 차량 위치 업데이트 DTO 타입
export type UpdateVehicleLocationDto = Partial<
  Omit<VehicleLocation, 'id' | 'vehicleId' | 'licensePlate' | 'createdAt' | 'updatedAt'>
>;

File: backend/services/fleet_api/lib/models/vehicle.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class VehicleStatus(str, Enum):
    """차량 상태 열거형"""

    AVAILABLE = "available"  # 이용 가능
    IN_USE = "in_use"  # 사용 중
    MAINTENANCE = "maintenance"  # 정비 중
    OUT_OF_SERVICE = "out_of_service"  # 운행 불가


class VehicleType(str, Enum):
    """차량 유형 열거형"""

    SEDAN = "sedan"  # 승용차
    SUV = "suv"  # SUV
    TRUCK = "truck"  # 트럭
    VAN = "van"  # 밴
    MOTORCYCLE = "motorcycle"  # 오토바이
    BUS = "bus"  # 버스
    OTHER = "other"  # 기타


class VehicleBase(BaseModel):
    """차량 기본 모델"""

    make: str = Field(..., description="제조사")
    model: str = Field(..., description="모델명")
    year: int = Field(..., description="연식", ge=1900, le=datetime.now().year + 1)
    vehicle_type: str = Field(..., description="차량 유형 (sedan, suv, truck 등)")
    license_plate: str = Field(..., description="번호판")
    vin: str = Field(..., description="차대번호")
    color: Optional[str] = Field(None, description="색상")
    mileage: int = Field(0, description="주행거리(km)", ge=0)
    fuel_type: Optional[str] = Field(None, description="연료 유형")
    transmission: Optional[str] = Field(None, description="변속기 유형")
    engine: Optional[str] = Field(None, description="엔진 정보")
    features: List[str] = Field([], description="차량 특징/옵션")
    notes: Optional[str] = Field(None, description="메모")
    status: str = Field(
        "available", description="상태 (available, in_use, maintenance, out_of_service)"
    )


class VehicleCreate(VehicleBase):
    """차량 생성 모델"""

    organization_id: str = Field(..., description="소속 조직 ID")
    purchase_date: Optional[datetime] = Field(None, description="구매일")
    purchase_price: Optional[float] = Field(None, description="구매 가격")
    registration_date: Optional[datetime] = Field(None, description="등록일")
    insurance_info: Optional[Dict[str, Any]] = Field(None, description="보험 정보")
    inspection_date: Optional[datetime] = Field(None, description="검사일")
    next_inspection_date: Optional[datetime] = Field(None, description="다음 검사일")
    is_active: bool = Field(True, description="활성 상태 여부")


class VehicleUpdate(BaseModel):
    """차량 업데이트 모델"""

    make: Optional[str] = None
    model: Optional[str] = None
    year: Optional[int] = None
    vehicle_type: Optional[str] = None
    license_plate: Optional[str] = None
    vin: Optional[str] = None
    color: Optional[str] = None
    mileage: Optional[int] = None
    fuel_type: Optional[str] = None
    transmission: Optional[str] = None
    engine: Optional[str] = None
    features: Optional[List[str]] = None
    notes: Optional[str] = None
    status: Optional[str] = None
    organization_id: Optional[str] = None
    purchase_date: Optional[datetime] = None
    purchase_price: Optional[float] = None
    registration_date: Optional[datetime] = None
    insurance_info: Optional[Dict[str, Any]] = None
    inspection_date: Optional[datetime] = None
    next_inspection_date: Optional[datetime] = None
    is_active: Optional[bool] = None


class VehicleResponse(VehicleBase):
    """차량 응답 모델"""

    id: str
    organization_id: str
    purchase_date: Optional[datetime] = None
    purchase_price: Optional[float] = None
    registration_date: Optional[datetime] = None
    insurance_info: Optional[Dict[str, Any]] = None
    inspection_date: Optional[datetime] = None
    next_inspection_date: Optional[datetime] = None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class VehicleListResponse(BaseModel):
    """차량 목록 응답"""

    items: List[VehicleResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
